util/expensivequery/expensivequery.go

```diff
@@ -57,10 +59,23 @@ func (eqh *Handle) Run() {
 		case <-ticker.C:
 			processInfo := sm.ShowProcessList()
 			for _, info := range processInfo {
+				if info.CurTxnStartTS != 0 {
+					txnCostTime := time.Since(info.CurTxnCreateTime)
+					if txnCostTime >= time.Second*time.Duration(txnThreshold) {
+						if info.StmtCtx.InRestrictedSQL {
+							metrics.OngoingTxnDurationHistogram.WithLabelValues(metrics.LblInternal).Observe(txnCostTime.Seconds())
+						} else {
+							metrics.OngoingTxnDurationHistogram.WithLabelValues(metrics.LblGeneral).Observe(txnCostTime.Seconds())
+						}
```

Comment:
It might be a little expensive for observing all long-running txns every `tickInterval` (100ms). How about init metrics with type label before for loop and observe every 15s (the default scrape interval of prometheus). eg:
```
ongoingTxnDurationHistogramInternal := metrics.OngoingTxnDurationHistogram.WithLabelValues(metrics.LblInternal)
ongoingTxnDurationHistogramGeneral := metrics.OngoingTxnDurationHistogram.WithLabelValues(metrics.LblGeneral)
lastOberveTime := time.Time{}
for {
  select <-ticker.C:
    needObserve := false
    if now := time.Now(); now.Since(lastOberveTime) > 15*time.Second {
      needObserve = true
      lastOberveTime = now
    }
    for info in sm.ShowProcessList() {
      if txnCostTime >= time.Second*time.Duration(txnThreshold) {
        if needObserve {
          if info.IsInternal {
            ongoingTxnDurationHistogramInternal.Observe(txnCostTime.Seconds())
          } else {
            ongoingTxnDurationHistogramGeneral.Observe(txnCostTime.Seconds())
          }
        }
        ...
      }
    } 
}```

Reply:
great, done.